### PR TITLE
Detect WP and skip the area on users request

### PR DESCRIPTION
**Pull request by SergiiDmytruk**
A simple test can look like this:
```
truncate --size=16M rom
./flashrom -p dummy:emulate=W25Q128FV,spi_status=0x00ec,hwwp=yes --write rom
```

### CODE REVIEW

**Comment by macpijan** [comment]
oh, we certainly do not want github bot close our PRs here automatically

File: Documentation/heads-and-wp.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+Partial updates can produce an unbootable image if old bootblock doesn't work
+with a more recent version of `coreboot`. This can be manifested in various
+ways ranging from old bootblock not being able to find new romstage to system
+booting successfully, but data in `coreboot` tables being mangled or incomplete.
+
+The incompatibilities might happen when switching version of firmware or when
+using the same version with a slightly different configuration.
+
+Another thing that can potentially cause trouble is CBFS layout. When bootblock
+is part of CBFS and doesn't have a fixed, it can also change its location when
```
Comment (macpijan) [inline Documentation/heads-and-wp.md]: a fixed ... ?Somethin's missing?

File: Documentation/heads-and-wp.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+it depends on its size. If newer bootblock is smaller such that an old range now
+covers bootblock and some other file, this new file won't be updated either
+potentially corrupting the image. Luckily, when bootblock is the last file it's
+normally preceded by a significant amount of empty space, which won't this
```
Comment (macpijan) [inline Documentation/heads-and-wp.md]: won't ... this <- I have a feeling that some verb might be missing here as well

File: Documentation/heads-and-wp.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
+## Updating Heads
+
+Whether Heads is to be updated in full or only in part makes no difference for
+the choice of the flash image, it's always the same full flash image.
+
+### Updating Heads after the bootblock has been locked
+
+When the patched `flashrom` detects an attempt to flash a write protected chip,
+it prints a warning informing the user about potential danger of proceeding and
+suggest to use `--skip-wp-area` parameter to perform the operation.
```
Comment (krystian-hebel) [inline Documentation/heads-and-wp.md]:
```suggestion
suggests to use `--skip-wp-area` parameter to perform the operation.
```

File: Documentation/heads-and-wp.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
+## Updating Heads
```
Comment (krystian-hebel) [inline Documentation/heads-and-wp.md:1]: Is this specific to Heads only or can it be applied to other firmware images as well?

File: cli_classic.c [hunk 5]
```diff
@@ -941,6 +948,7 @@ int main(int argc, char *argv[])
 
 	fill_flash = &flashes[0];
 
+
```
Comment (krystian-hebel) [inline cli_classic.c]: Not needed

File: Documentation/heads-and-wp.md (hunk 4 above)
Comment (SergiiDmytruk) [reply Documentation/heads-and-wp.md:1]: It's generally applicable, but I'm not sure where this documentation will end up, it might be Heads' wiki.

**Comment by macpijan** [comment]
@krystian-hebel Any more comments to that?

Generally, it looks good to me.

@tlaurion Do you have any more comments to that? I know you have mentioned somewhere else that in the end it might be difficult to apply this approach to heads.

But this MR was focusing on the goal:

```
Have ﬂashrom deal properly with that (WP) region when upgrading full ROM internally
```

which seems to be covered.

**Review by krystian-hebel — APPROVED** [review]
Looks good to me. Depending on where this will land eventually, notes column in `chips.md` may require redacting.

**Comment by tlaurion** [comment]
@macpijan @krystian-hebel 

I can only speak from past experience here, but from current documentation, it seems that the user would receive a warning saying that bootblock region is going to have some changes applied PRIOR of applying them, requiring the user to adapt its command line arguments to apply explicitly do the change.

As for Heads, or any internal upgrade (think FWUPD) I'm still unsure how that would be implemented.

The ideal would be for the user to not do the obviously stupid move of flashing the whole SPI if bootblock region was not expected to change from internal flash upgrade. 

And I guess some wrapper telling to the user to stay on a same coreboot version branch from GUI output, and/or to externally flash the firmware image on the platform.

**Comment by tlaurion** [comment]
@macpijan @krystian-hebel ping on  https://github.com/Dasharo/flashrom/pull/6#issuecomment-1261120655

Also, looking to find chips.md which seems to have been lost in master/upstream; https://github.com/Dasharo/flashrom/commit/94083d038020e749a1712074a5f521d3ed78216b

Was that intended?

**Pull request merged**